---
title: "clinvaR Vignette"
output:
  html_document: default
  html_notebook: default
---

```{r setup, include = F}
knitr::opts_knit$set(root.dir = ".");
knitr::opts_chunk$set(echo = T, eval = T, cache = T, warning = F, message = F)
```

This is a vignette intended to introduce new users to [clinvaR](https://github.com/jamesdiao/clinvaR). 

The first step is to have a list of genes that you are interested in. 
The easiest way is to directly save your genes of interest in gene_panel as a character vector (all caps):

```{r}
pkg_list <- c("knitr","pander","ggplot2","clinvaR",
              "tibble","curl","tidyr","dplyr", "stringr")
installed <- pkg_list %in% installed.packages()[,"Package"]
if (!all(installed))
  install.packages(pkg_list[!installed])
sapply(pkg_list, require, character.only = T)
```


```{r}
gene_panel <- c('GENE1', 'GENE2', 'GENE3')
```

Another way is to save your gene list in a text file (1 gene per line) and import it. 

The MacArthur Lab has assembled a compilation of gene lists ([https://github.com/macarthur-lab/gene_lists](https://github.com/macarthur-lab/gene_lists)) that may be a useful starting point. We have included all sets with abbreviated descriptions. Please visit the GitHub repo for citation information. 

| List | File Name | Count | Description |
| ---------------- | ------- | ------- | ------------------------------------------------------- | 
| [Universe](lists/universe.tsv) | universe.tsv | 18,991 | Approved symbols for 18,991 protein-coding genes according to HGNC. All other lists are subsets. |
| [FDA-approved drug targets](lists/fda_approved_drug_targets.tsv) | fda_approved_drug_targets.tsv | 286 | Genes whose protein products are known to be the mechanistic targets of FDA-approved drugs. |
| [Drug targets by Nelson et al 2012](lists/drug_targets_nelson.tsv) | drug_targets_nelson.tsv | 201 | Drug targets according to Nelson et al 2012. |
| [Autosomal dominant genes by Blekhman et al 2008](lists/blekhman_ad.tsv) | blekhman_ad.tsv | 307 | OMIM genes that follow autosomal dominant inheritance according to Molly Przeworski's group. |
| [Autosomal dominant genes by Berg et al 2013](lists/berg_ad.tsv) | berg_ad.tsv | 631 | OMIM genes that follow autosomal dominant inheritance according to Berg et al, 2013. |
| [Autosomal recessive genes by Blekhman et al 2008](lists/blekhman_ar.tsv) | blekhman_ar.tsv | 529 | OMIM genes that follow autosomal recessive inheritance according to Molly Przeworski's group. |
| [Autosomal recessive genes by Berg et al 2013](lists/berg_ar.tsv) | berg_ar.tsv | 1073 | OMIM disease genes (as of June 2011) deemed to follow autosomal recessive inheritance according Berg et al, 2013. |
| [X-linked genes by Blekhman et al 2008](lists/blekhman_x.tsv) | blekhman_x.tsv | 66 | OMIM disease genes deemed to follow X-linked inheritance (dominant/recessive not specified) according to extensive manual curation by Molly Przeworski's group. | 
| [X-linked recessive genes by Berg et al 2013](lists/berg_xr.tsv) | berg_xr.tsv | 102 | OMIM disease genes (as of June 2011) deemed to follow X-linked recessive inheritance according Berg et al, 2013. |
| [X-linked dominant genes by Berg et al 2013](lists/berg_xd.tsv) | berg_xd.tsv | 34 | OMIM disease genes (as of June 2011) deemed to follow X-linked dominant inheritance according Berg et al, 2013. | 
| [X-linked ClinVar genes](lists/x-linked_clinvar.tsv) | x-linked_clinvar.tsv | 61 | X chromosome genes in the August 6, 2015 ClinVar release that have at least 3 reportedly pathogenic, non-conflicted variants in ClinVar with at least one submitter other than OMIM or GeneReviews.|
| [All dominant genes](lists/all_ad.tsv) | all_ad.tsv | 709 | Currently the union of the Berg and Blekhman dominant lists, may add more lists later. |
| [All recessive genes](lists/all_ar.tsv) | all_ar.tsv | 1183 | Currently the union of the Berg and Blekhman recessive lists, may add more lists later. | 
| [Essential in culture](lists/core_essentials_hart.tsv) | core_essentials_hart.tsv | 285 | Genes deemed essential in multiple cultured cell lines based on shRNA screen data |
| [Essential in mice](lists/mgi_essential.tsv) | mgi_essential.tsv | 2,454 | Genes where homozygous knockout in mice results in pre-, peri- or post-natal lethality. |
| [Genes nearest to GWAS peaks](lists/gwascatalog.tsv) | gwascatalog.tsv | 3,762 | Closest gene 3' and 5' of GWAS hits in the NHGRI GWAS catalog as of Feb 9, 2015 |
| [DNA Repair Genes, WoodRD](lists/DRG_WoodRD.tsv) | DRG_WoodRD.tsv | 178 | An updated inventory of human DNA repair genes. |
| [DNA Repair Genes, KangJ](lists/DRG_KangJ.tsv) | DRG_KangJ.tsv | 151 | 151 DNA repair genes from DNA repair pathways: ATM, BER, FA/HR, MMR, NHEJ, NER, TLS, XLR, RECQ, and other. | 
| [ClinGen haploinsufficient genes](lists/clingen_level3_genes_2015_02_27.tsv) | clingen_level3_genes_2015_02_27.tsv | 221 | Genes with sufficient evidence for dosage pathogenicity (level 3) as determined by the ClinGen Dosage Sensitivity Map |
| [Olfactory receptors](lists/olfactory_receptors.tsv) | olfactory_receptors.tsv | 371 | Olfactory receptors from the Mainland 2015's [data release](http://files.figshare.com/1816348/Receptors.tsv) |
| [Genes with any disease association reported in ClinVar](lists/clinvar_path_likelypath.tsv) | clinvar_path_likelypath.tsv | 3078 | All gene symbols for which there is at least one variant with an assertion of pathogenic or likely pathogenic in ClinVar. | 
| [Kinases](lists/kinases.tsv) | kinases.tsv | 351 | From UniProt's [pkinfam list](http://www.uniprot.org/docs/pkinfam) |
| [GPCRs](lists/gpcr.tsv) | gpcr.tsv | 1705 | GPCR list from [guidetopharmacology.org](http://www.guidetopharmacology.org/GRAC/GPCRListForward?class=A) | 
| [Natural product targets](lists/natural_product_targets.tsv) | natural_product_targets.tsv | 37 | List of hand-curated targets of natural products |
| [BROCA - Cancer Risk Panel](lists/BROCA_Cancer_Risk_Panel.tsv) | BROCA_Cancer_Risk_Panel.tsv | 66 | Suspected hereditary cancer predisposition, with a focus on breast or ovarian cancers. May co-occur with other cancer types (such as colorectal, endometrial, pancreatic, endocrine, or melanoma). |

```{r}
import_gene_file <- function(file) {
  read.table(sprintf('gene_lists/lists/%s', file)) %>% 
             unlist() %>% unique() %>% as.character()
}
gene_panel <- import_gene_file('x-linked_clinvar.tsv')
gene_panel[1:5]
```

The last way is to search OMIM for related genes. 
```{r}
#omim_table <- read.table('clinvaR/storage/gene_lists/other_data/omim.full.tsv', 
#                         sep = '\t', fill = T, header = T)
#search_terms <- c("Cardiomyopathy")
```

Then, you can download variants of interest from 1000 Genomes

```{r}
download_output <- download_1000g(genes = gene_panel)
final_vcf <- import_file_1000g(genes = gene_panel)
```

This must be merged with some version of ClinVar. This can be done either using the latest saved version (July 5, 2017), or by specifying a date from the list. 

```{r}
require(stringr)
setwd('/Users/jamesdiao/Documents/Kohane_Lab/clinvaR/inst/extdata')
get_date_list <- function() {
  clinvar_reports <- system("ls Archive_Tables", intern = T)
  clinvar_reports <- clinvar_reports[grep(".tsv",clinvar_reports)]
  do.call("rbind", lapply(clinvar_reports, function(tsv) {
    read.table(file = sprintf("Archive_Tables/%s", tsv), sep = "\t", 
               header = T, stringsAsFactors = F)
  })) -> archive
  regmatches(archive$Name,regexpr("^File:clinvar_(20.{6})\\.vcf\\.gz$",archive$Name)) %>% 
    str_extract("20.{6}") %>% as.Date(format = "%Y%m%d") %>% unique()
}
all_dates <- get_date_list()

get_dates <- function(date_list, range) {
  date_list <- date_list %>% as.Date()
  if (!exists("all_dates"))
    all_dates <- get_date_list()
  if (range) {
    keep <- between(all_dates, min(date_list), max(date_list))
  } else {
    keep <- all_dates %in% date_list
  }
  all_dates[keep] %>% as.character()
}
clinvar <- get_test_clinvar()
```

Then, we merge clinvar with the genes of interest. We can also slice clinvar itself to collect only
relevant entries. 

```{r}
sapply(gene_panel, gene, function(x) {
  
})
clinvar$CHROM %>% paste0(str_pad(clinvar$POS, 10, side = 'left', pad = "0")) %>% as.numeric %>% 
  between(start, stop) -> filter_condition
clinvar <- filter(clinvar, filter_condition)
```



```{r, eval = F, echo = F}
# One entry per gene. Start/end refer to transcription regions. When multiple exist, the longest region is taken. 
query <- connect_UCSC()
UCSC <- query('select * from refGene') %>% select(gene = name2, name, chrom, start = txStart, end = txEnd)
chr <- gsub("chr(.*)", "\\1", UCSC$chrom)
UCSC$chrom <- chr %>% replace(chr=="X", 23) %>% replace(chr=="Y", 24)
UCSC <- UCSC %>% filter(chr %in% c(1:22, "X","Y")) %>% 
  mutate(chrom = as.integer(chrom)) %>% 
  arrange(desc(end-start), gene, start) %>% 
  filter(!duplicated(gene)) %>% 
  arrange(chrom, gene)
```
