<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="James Diao" />

<meta name="date" content="2017-07-29" />

<title>clinvaR vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">clinvaR vignette</h1>
<h4 class="author"><em>James Diao</em></h4>
<h4 class="date"><em>2017-07-29</em></h4>



<p>This is a vignette intended to introduce new users to <code>clinvaR</code> (<a href="https://github.com/jamesdiao/clinvaR" class="uri">https://github.com/jamesdiao/clinvaR</a>).</p>
<div id="purpose" class="section level2">
<h2>Purpose</h2>
<p>For <code>clinvaR</code>, the first goal is to go from genes to a vcf of related variants, along with classifications from any version of ClinVar.<br />
This involves 3 steps:</p>
<ol style="list-style-type: decimal">
<li>Select a list of genes that you are interested in. clinvaR makes it easy to download and import relevant variants from 1KG.<br />
</li>
<li>Select a version of ClinVar to assign pathogenicity labels with. All previous ClinVar VCFs are stored as binary files and can be retrieved with a “closest date” function.<br />
</li>
<li>Select an analysis method. This part is not ready yet.</li>
</ol>
</div>
<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>clinvaR has a number of file dependencies that are generally useful for these types of projects. I have attached them here.</p>
<pre><code>## [1] remotes pander  ggplot2 tibble  tidyr   dplyr   stringr</code></pre>
</div>
<div id="select-a-list-of-genes" class="section level2">
<h2>Select a List of Genes</h2>
<p>The first step is to have a list of genes that you are interested in.</p>
<div id="method-1-direct-assignment" class="section level3">
<h3>Method 1: Direct assignment</h3>
<p>The easiest way is to directly save your genes of interest in gene_panel as a character vector (all caps). We have taken the example of the 20 genes on the Laboratory of Molecular Medicine’s gene testing panel for hypertrophic cardiomyopathy.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_panel &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ACTC1&quot;</span>, <span class="st">&quot;ACTN2&quot;</span>, <span class="st">&quot;CSRP3&quot;</span>, <span class="st">&quot;GLA&quot;</span>, <span class="st">&quot;LAMP2&quot;</span>, <span class="st">&quot;MYBPC3&quot;</span>, <span class="st">&quot;MYH7&quot;</span>, 
<span class="st">&quot;MYL2&quot;</span>, <span class="st">&quot;MYL3&quot;</span>, <span class="st">&quot;MYOZ2&quot;</span>, <span class="st">&quot;NEXN&quot;</span>, <span class="st">&quot;PLN&quot;</span>, <span class="st">&quot;PRKAG2&quot;</span>, <span class="st">&quot;PTPN11&quot;</span>, <span class="st">&quot;RAF1&quot;</span>, 
<span class="st">&quot;TNNC1&quot;</span>, <span class="st">&quot;TNNI3&quot;</span>, <span class="st">&quot;TNNT2&quot;</span>, <span class="st">&quot;TPM1&quot;</span>, <span class="st">&quot;TTR&quot;</span>)</code></pre></div>
</div>
<div id="method-2-import-from-text-file" class="section level3">
<h3>Method 2: Import from text file</h3>
<div id="from-own-file" class="section level4">
<h4>From own file</h4>
<p>Another way is to save your gene list in a text file (1 gene per line) and import it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata/MacArthur_Gene_Lists/lists/lmm_hcm.tsv&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;clinvaR&quot;</span>)
<span class="kw">print</span>(path)</code></pre></div>
<pre><code>## [1] &quot;/Users/jamesdiao/Documents/Kohane_Lab/clinvaR/inst/extdata/MacArthur_Gene_Lists/lists/lmm_hcm.tsv&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_panel &lt;-<span class="st"> </span><span class="kw">get_genes</span>(path)
<span class="kw">print</span>(gene_panel)</code></pre></div>
<pre><code>##  [1] &quot;ACTC1&quot;  &quot;ACTN2&quot;  &quot;CSRP3&quot;  &quot;GLA&quot;    &quot;LAMP2&quot;  &quot;MYBPC3&quot; &quot;MYH7&quot;  
##  [8] &quot;MYL2&quot;   &quot;MYL3&quot;   &quot;MYOZ2&quot;  &quot;NEXN&quot;   &quot;PLN&quot;    &quot;PRKAG2&quot; &quot;PTPN11&quot;
## [15] &quot;RAF1&quot;   &quot;TNNC1&quot;  &quot;TNNI3&quot;  &quot;TNNT2&quot;  &quot;TPM1&quot;   &quot;TTR&quot;</code></pre>
</div>
<div id="from-clinvar-stored-lists" class="section level4">
<h4>From clinvaR stored lists</h4>
<p>The MacArthur Lab has assembled a compilation of gene lists (<a href="https://github.com/macarthur-lab/gene_lists" class="uri">https://github.com/macarthur-lab/gene_lists</a>) that may be a useful starting point. We have included all sets (and a few of our own) with abbreviated descriptions. Please visit the GitHub repo for citation information.</p>
<table>
<colgroup>
<col width="19%"></col>
<col width="8%"></col>
<col width="8%"></col>
<col width="62%"></col>
</colgroup>
<thead>
<tr class="header">
<th>List</th>
<th>File Name</th>
<th>Count</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Universe</td>
<td>universe.tsv</td>
<td>18,991</td>
<td>Approved symbols for 18,991 protein-coding genes according to HGNC. All other lists are subsets.</td>
</tr>
<tr class="even">
<td>Hypertrophic cardiomyopathy</td>
<td>lmm_hcm.tsv</td>
<td>20</td>
<td>Gene testing panel for hypertrophic cardiomyopathy from the Laboratory of Molecular Medicine</td>
</tr>
<tr class="odd">
<td>ACMG secondary findings recommendations v2.0</td>
<td>acmg_59.tsv</td>
<td>59</td>
<td>The American College of Medical Genetics and Genomics has recommended that sequencing laboratories seek and report secondary findings in a minimum list of 59 genes.</td>
</tr>
<tr class="even">
<td>BROCA - Cancer Risk Panel</td>
<td>BROCA_Cancer_Risk_Panel.tsv</td>
<td>66</td>
<td>Suspected hereditary cancer predisposition, with a focus on breast or ovarian cancers. May co-occur with other cancer types (such as colorectal, endometrial, pancreatic, endocrine, or melanoma).</td>
</tr>
<tr class="odd">
<td>DNA Repair Genes, KangJ</td>
<td>DRG_KangJ.tsv</td>
<td>151</td>
<td>151 DNA repair genes from DNA repair pathways: ATM, BER, FA/HR, MMR, NHEJ, NER, TLS, XLR, RECQ, and other.</td>
</tr>
<tr class="even">
<td>ClinGen haploinsufficient genes</td>
<td>clingen_level3_genes_2015_02_27.tsv</td>
<td>221</td>
<td>Genes with sufficient evidence for dosage pathogenicity (level 3) as determined by the ClinGen Dosage Sensitivity Map</td>
</tr>
<tr class="odd">
<td>Genes with any disease association reported in ClinVar</td>
<td>clinvar_path_likelypath.tsv</td>
<td>3078</td>
<td>All gene symbols for which there is at least one variant with an assertion of pathogenic or likely pathogenic in ClinVar.</td>
</tr>
<tr class="even">
<td>FDA-approved drug targets</td>
<td>fda_approved_drug_targets.tsv</td>
<td>286</td>
<td>Genes whose protein products are known to be the mechanistic targets of FDA-approved drugs.</td>
</tr>
<tr class="odd">
<td>Drug targets by Nelson et al 2012</td>
<td>drug_targets_nelson.tsv</td>
<td>201</td>
<td>Drug targets according to Nelson et al 2012.</td>
</tr>
<tr class="even">
<td>X-linked ClinVar genes</td>
<td>x-linked_clinvar.tsv</td>
<td>61</td>
<td>X chromosome genes in the August 6, 2015 ClinVar release that have at least 3 reportedly pathogenic, non-conflicted variants in ClinVar with at least one submitter other than OMIM or GeneReviews.</td>
</tr>
<tr class="odd">
<td>All dominant genes</td>
<td>all_ad.tsv</td>
<td>709</td>
<td>Currently the union of the Berg and Blekhman autosomal dominant OMIM gene lists, may add more lists later.</td>
</tr>
<tr class="even">
<td>All recessive genes</td>
<td>all_ar.tsv</td>
<td>1183</td>
<td>Currently the union of the Berg and Blekhman autosomal recessive OMIM gene lists, may add more lists later.</td>
</tr>
<tr class="odd">
<td>Essential in culture</td>
<td>core_essentials_hart.tsv</td>
<td>285</td>
<td>Genes deemed essential in multiple cultured cell lines based on shRNA screen data</td>
</tr>
<tr class="even">
<td>Essential in mice</td>
<td>mgi_essential.tsv</td>
<td>2,454</td>
<td>Genes where homozygous knockout in mice results in pre-, peri- or post-natal lethality.</td>
</tr>
<tr class="odd">
<td>Genes nearest to GWAS peaks</td>
<td>gwascatalog.tsv</td>
<td>3,762</td>
<td>Closest gene 3’ and 5’ of GWAS hits in the NHGRI GWAS catalog as of Feb 9, 2015</td>
</tr>
<tr class="even">
<td>Kinases</td>
<td>kinases.tsv</td>
<td>351</td>
<td>From UniProt’s <a href="http://www.uniprot.org/docs/pkinfam">pkinfam list</a></td>
</tr>
<tr class="odd">
<td>G-protein-coupled receptors</td>
<td>gpcr.tsv</td>
<td>1705</td>
<td>GPCR list from <a href="http://www.guidetopharmacology.org/GRAC/GPCRListForward?class=A">guidetopharmacology.org</a></td>
</tr>
<tr class="even">
<td>Natural product targets</td>
<td>natural_product_targets.tsv</td>
<td>37</td>
<td>List of hand-curated targets of natural products</td>
</tr>
</tbody>
</table>
<p>These are easily imported by directly referring to the file name.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_panel &lt;-<span class="st"> </span><span class="kw">get_genes</span>(<span class="st">'lmm_hcm.tsv'</span>)
<span class="kw">print</span>(gene_panel)</code></pre></div>
<pre><code>##  [1] &quot;ACTC1&quot;  &quot;ACTN2&quot;  &quot;CSRP3&quot;  &quot;GLA&quot;    &quot;LAMP2&quot;  &quot;MYBPC3&quot; &quot;MYH7&quot;  
##  [8] &quot;MYL2&quot;   &quot;MYL3&quot;   &quot;MYOZ2&quot;  &quot;NEXN&quot;   &quot;PLN&quot;    &quot;PRKAG2&quot; &quot;PTPN11&quot;
## [15] &quot;RAF1&quot;   &quot;TNNC1&quot;  &quot;TNNI3&quot;  &quot;TNNT2&quot;  &quot;TPM1&quot;   &quot;TTR&quot;</code></pre>
<p>The last way is to search OMIM for related genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#omim_table &lt;- read.table('clinvaR/storage/gene_lists/other_data/omim.full.tsv', </span>
<span class="co">#                         sep = '\t', fill = T, header = T)</span>
<span class="co">#search_terms &lt;- c(&quot;Cardiomyopathy&quot;)</span></code></pre></div>
<p>Then, you can download variants of interest from 1000 Genomes</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#download_output &lt;- download_1000g(genes = gene_panel)</span>
<span class="co">#final_vcf &lt;- import_file_1000g(genes = gene_panel)</span></code></pre></div>
</div>
</div>
</div>
<div id="import-clinvar-vcf" class="section level2">
<h2>Import ClinVar VCF</h2>
<p>This must be merged with some version of ClinVar. This can be done either using the latest saved version (July 5, 2017), or by specifying a date from the list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">get_date_list</span>()</code></pre></div>
<pre><code>##  [1] &quot;2012-06-16&quot; &quot;2012-10-26&quot; &quot;2012-11-05&quot; &quot;2012-12-31&quot; &quot;2013-01-14&quot;
##  [6] &quot;2013-01-18&quot; &quot;2013-02-26&quot; &quot;2013-05-06&quot; &quot;2013-08-08&quot; &quot;2013-09-05&quot;
## [11] &quot;2013-09-30&quot; &quot;2013-11-05&quot; &quot;2013-12-03&quot; &quot;2013-12-30&quot; &quot;2014-02-11&quot;
## [16] &quot;2014-03-03&quot; &quot;2014-04-01&quot; &quot;2014-04-30&quot; &quot;2014-06-04&quot; &quot;2014-07-02&quot;
## [21] &quot;2014-08-07&quot; &quot;2014-09-02&quot; &quot;2014-09-29&quot; &quot;2014-11-05&quot; &quot;2014-12-02&quot;
## [26] &quot;2015-01-06&quot; &quot;2015-02-03&quot; &quot;2015-03-05&quot; &quot;2015-03-30&quot; &quot;2015-05-04&quot;
## [31] &quot;2015-06-03&quot; &quot;2015-06-29&quot; &quot;2015-08-04&quot; &quot;2015-09-01&quot; &quot;2015-09-29&quot;
## [36] &quot;2015-11-02&quot; &quot;2015-12-01&quot; &quot;2016-01-04&quot; &quot;2016-02-03&quot; &quot;2016-03-02&quot;
## [41] &quot;2016-04-05&quot; &quot;2016-05-02&quot; &quot;2016-05-31&quot; &quot;2016-07-05&quot; &quot;2016-08-02&quot;
## [46] &quot;2016-08-31&quot; &quot;2016-10-03&quot; &quot;2016-11-01&quot; &quot;2016-11-28&quot; &quot;2017-01-04&quot;
## [51] &quot;2017-01-30&quot; &quot;2017-02-28&quot; &quot;2017-04-04&quot; &quot;2017-05-01&quot; &quot;2017-05-30&quot;
## [56] &quot;2017-07-05&quot;</code></pre>
<p>Use <code>get_clinvar()</code> to collect from a certain date (if date is not present, the closest date is used).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clinvar &lt;-<span class="st"> </span><span class="kw">get_clinvar</span>(<span class="st">&quot;2017-07-21&quot;</span>)
clinvar[<span class="dv">1</span>:<span class="dv">4</span>,] <span class="co">#%&gt;% select(-VAR_ID, -CLNDBN)</span></code></pre></div>
<pre><code>##         VAR_ID  GENE CHROM    POS REF ALT          ID CLNSIG
## 1 1_949523_C_T ISG15     1 949523   C   T rs786201005      5
## 2 1_949608_G_A ISG15     1 949608   G   A      rs1921      2
## 3 1_949739_G_T ISG15     1 949739   G   T rs672601312      5
## 4 1_955563_G_C  AGRN     1 955563   G   C rs539283387      3
##                                                 CLNDBN
## 1 Immunodeficiency_38_with_basal_ganglia_calcification
## 2                                        not_specified
## 3 Immunodeficiency_38_with_basal_ganglia_calcification
## 4                                        not_specified
##   pathogenic_incl_conflicts pathogenic_no_conflicts
## 1                      TRUE                    TRUE
## 2                     FALSE                   FALSE
## 3                      TRUE                    TRUE
## 4                     FALSE                   FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clinvar &lt;-<span class="st"> </span><span class="kw">get_clinvar</span>()
clinvar[<span class="dv">1</span>:<span class="dv">4</span>,] <span class="co">#%&gt;% select(-VAR_ID, -CLNDBN)</span></code></pre></div>
<pre><code>##         VAR_ID  GENE CHROM    POS REF ALT          ID CLNSIG
## 1 1_949523_C_T ISG15     1 949523   C   T rs786201005      5
## 2 1_949608_G_A ISG15     1 949608   G   A      rs1921      2
## 3 1_949739_G_T ISG15     1 949739   G   T rs672601312      5
## 4 1_955563_G_C  AGRN     1 955563   G   C rs539283387      3
##                                                 CLNDBN
## 1 Immunodeficiency_38_with_basal_ganglia_calcification
## 2                                        not_specified
## 3 Immunodeficiency_38_with_basal_ganglia_calcification
## 4                                        not_specified
##   pathogenic_incl_conflicts pathogenic_no_conflicts
## 1                      TRUE                    TRUE
## 2                     FALSE                   FALSE
## 3                      TRUE                    TRUE
## 4                     FALSE                   FALSE</code></pre>
<p>Then, we merge clinvar with the genes of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#merged_1000g &lt;- merge_clinvar_1000g(clinvar = get_clinvar(Sys.Date()), </span>
<span class="co">#                    vcf = import_file_1000g(gene_panel))</span>
<span class="co">#merged_1000g[1:5,1:5]</span></code></pre></div>
<p>We can also slice clinvar itself to collect only relevant entries.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#clinvar$CHROM %&gt;% paste0(str_pad(clinvar$POS, 10, side = 'left', pad = &quot;0&quot;)) %&gt;% as.numeric %&gt;% </span>
<span class="co">#  between(start, stop) -&gt; filter_condition</span>
<span class="co">#clinvar &lt;- filter(clinvar, filter_condition)</span></code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
