#' Population frequencies over time
#'
#' This function computes the probability of having at least 1 variant  
#' in the gene list of interest (assuming independence) for a range of ClinVar files
#' 
#' @usage freq_over_time(vcf)
#' @param vcf data.frame; clinvaR-processed vcf from import_file_1000g()
#' @examples 
#' freq_over_time(import_file_1000g())
#' @export

freq_over_time <- function(vcf, dates, verbose) {
  if (missing(vcf)) 
    stop('Error: freq_over_time() requires a VCF generated by import_file_1000g()')
  if (missing(dates))
    dates <- get_date_list()
  if (missing(verbose))
    verbose <- TRUE
  freqs <- lapply(dates, function(dat) {
    if (verbose)
      print(sprintf("%s [%s/%s]", dat, which(dat == dates), length(dates)), quote = F)
    merged_vcf <- annotate_1000g(clinvar = get_clinvar(dat), vcf = vcf) 
    return(1-prod(1-merged_vcf$AF_1000G))
  }) %>% unlist %>% setNames(dates)
  class(freqs) <- c("freq_dates", class(freqs))
  plot(freqs)
  return(freqs)
}


#' Plot Aggregate Population Frequencies Over Time
#'
#' This function plots the aggregate population frequency generated by freq_over_time(). 
#'    
#' @usage plot(freqs)
#' @param freqs the named frequencies output by freq_over_time(). 
#' @export

plot.freq_dates <- function(freqs) {
  dates <- names(freqs) %>% as.Date()
  xlim <- as.Date(sprintf("%s-01-01", as.integer(range(format(dates,"%Y"))) + c(0,1)))
  plot(dates, freqs, type = 'o', xlim = xlim, #las = 2, xaxt = "n", xlab = "", 
       xlab = "Dates", ylab = "Frequency", main = "Aggregate Population Frequency")
  #minor <- seq(xlim[1], xlim[2], by = "12 months")
  #axis.Date(side = 1, dates, at = minor, format = "%b %Y", las = 2)
}


#' Aggregated penetrance over time
#'
#' This function computes the penetrance of a positive finding
#' in the gene list of interest (assuming independence) 
#' for a range of ClinVar files. Prevalence and Case Frequency values must be given. 
#' 
#' @usage penetrance_over_time(vcf)
#' @param vcf data.frame; clinvaR-processed vcf from import_file_1000g()
#' @examples 
#' penetrance_over_time(import_file_1000g())
#' @export

penetrance_over_time <- function(vcf, prevalence, case_freq, dates, verbose) {
  if (missing(vcf)) 
    stop('Error: penetrance_over_time() requires a VCF generated by import_file_1000g()')
  if (missing(dates))
    dates <- get_date_list()
  if (missing(prevalence))
    stop("Error: penetrance_over_time() requires a prevalence estimate (prevalence)")
  if (missing(case_freq))
    stop("Error: penetrance_over_time() requires a case allele frequency estimate (case_freq)")
  if (missing(verbose))
    verbose <- TRUE
  penetrance <- lapply(dates, function(dat) {
    if (verbose)
      print(sprintf("%s [%s/%s]", dat, which(dat == dates), length(dates)), quote = F)
    merged_vcf <- annotate_1000g(clinvar = get_clinvar(dat), vcf = vcf) 
    pop_freq <- 1-prod(1-merged_vcf$AF_1000G)
    penetrance <- prevalence*case_freq/pop_freq
  }) %>% unlist %>% setNames(dates)
  class(penetrance) <- c("penetrance_dates", class(penetrance))
  plot(penetrance)
  return(penetrance)
}


#' Plot Aggregated Penetrance Over Time
#'
#' This function plots the aggregated penetrance values 
#' generated by penetrance_over_time(). 
#'    
#' @usage plot(penetrance)
#' @param penetrance the named penetrance output by penetrance_over_time(). 
#' @export

plot.penetrance_dates <- function(penetrance) {
  dates <- names(penetrance) %>% as.Date()
  xlim <- as.Date(sprintf("%s-01-01", as.integer(range(format(dates,"%Y"))) + c(0,1)))
  plot(dates, penetrance, type = 'o', xlim = xlim, #las = 2, xaxt = "n", xlab = "", 
       xlab = "Dates", ylab = "Penetrance", main = "Aggregated Penetrance Over Time")
  #minor <- seq(xlim[1], xlim[2], by = "12 months")
  #axis.Date(side = 1, dates, at = minor, format = "%b %Y", las = 2)
}


#' Number of Submissions Over Time
#'
#' This function computes the number of unique submissions for point mutations
#' (identified by chromosome, position, reference, alternate) in ClinVar
#' 
#' @usage submissions_over_time(genes)
#' @param genes character; list of genes. 
#' @examples 
#' submissions_over_time(get_genes('lmm_hcm.tsv'))
#' @export

submissions_over_time <- function(genes, dates, verbose) {
  if (missing(genes)) {
    stop('Error: submissions_over_time() requires a list of genes')
  }
  if (missing(dates))
    dates <- get_date_list()
  if (missing(verbose))
    verbose <- TRUE
  submissions <- lapply(dates, function(dat) {
    if (verbose)
      print(sprintf("%s [%s/%s]", dat, which(dat == dates), length(dates)), quote = F)
    clinvar <- get_clinvar(dat)
    return(sum(clinvar$GENE %in% genes))
  }) %>% unlist %>% setNames(dates)
  class(submissions) <- c("submission_dates", class(submissions))
  plot(submissions)
  return(submissions)
}


#' Plot Submissions Over Time
#'
#' This function plots the aggregate population frequency generated by submissions_over_time(). 
#'    
#' @usage plot(submissions)
#' @param submissions the named frequencies output by submissions_over_time(). 
#' @export

plot.submission_dates <- function(submissions, ...) {
  dates <- names(submissions) %>% as.Date()
  plot(dates, submissions, type = 'o', 
       xlim = as.Date(sprintf("%s-01-01", as.integer(range(format(dates,"%Y"))) + c(0,1))),
       xlab = "Dates", ylab = "Submissions", main = "Submissions Over Time")
}
