
#' ClinVar Over Time Analyses
#'
#' This function computes aggregate frequency, number of related positions, 
#' number of related assertions, and (if given case allele frequency and prevalence)
#' aggregate penetrance across all stored ClinVar files.
#' Returns a list of 2 or 3 
#' 
#' @usage over_time(vcf)
#' over_time(genes)
#' @param vcf data.frame; clinvaR-processed vcf from import_file_1000g()
#' @param genes character; list of genes of interest. 
#' @examples 
#' over_time(import_file_1000g())
#' @export

over_time <- function(vcf, dates, verbose, plot.it, prevalence, case_freq) {
  map <- system.file("extdata/Supplementary_Files/phase3map.txt", package = "clinvaR") %>% 
    read.table(stringsAsFactors = F, header = T) %>% as.data.frame()
  sig_map <- c("Uncertain","Not Provided","Benign","Likely Benign","Likely Pathogenic","Pathogenic","Drug Response","Histocompatibility","Other") %>% 
    setNames(c(0, 1, 2, 3, 4, 5, 6, 7, 255))
  if (missing(vcf)) 
    stop('Error: over_time() requires an unannotated VCF input')
  if (!("plain_vcf" %in% class(vcf)) | ("annotated_vcf" %in% class(vcf)))
    stop('Error: over_time() requires an unannotated VCF input')
  if (missing(dates))
    dates <- get_date_list()
  if (missing(verbose))
    verbose <- TRUE
  if (missing(plot.it))
    plot.it <- TRUE
  genes <- vcf$GENE %>% unique
  ptm <- proc.time()
  output <- lapply(as.character(dates), function(dat) {
    if (verbose) {
      elapsed <- h_read((proc.time() - ptm)['elapsed'])
      print(sprintf("%s [%s/%s], Time Elapsed: %s", dat, 
                    which(dat == as.character(dates)), 
                    length(dates), elapsed), 
            quote = F)
    }
    use_clinvar <- get_clinvar(dat)
    positions <- sum(use_clinvar$GENE %in% genes)
    assertions <- sapply(use_clinvar$CLNSIG, length) %>% 
      aggregate(by = list(GENE = use_clinvar$GENE), sum) %>% 
      filter(GENE %in% genes) %>% select(x) %>% sum
    significances <- use_clinvar$CLNSIG %>% unlist %>% 
      factor(levels = c(0, 1, 2, 3, 4, 5, 6, 7, 255)) %>% table
    names(significances) <- sig_map[as.character(names(significances))]
    merged_vcf <- annotate_1000g(clinvar = use_clinvar, vcf = vcf) 
    input <- merged_vcf %>% select(grep("(HG)|(NA)", colnames(merged_vcf))) %>% colSums 
    input <- rbind(aggregate(input, by = list(population = map$super_pop), FUN = mean),
                   c("TOTAL", mean(input)))
    freqs <- input$x %>% as.numeric %>% setNames(input$population)
    return(list(freqs, positions, assertions, significances) %>% 
             setNames(c('Freqs','Positions','Assertions','Significances')))
    #return(1-prod(1-merged_vcf$AF_1000G)^2)
  }) %>% setNames(as.character(dates))
  freqs <- sapply(output, function(item) item[['Freqs']] ) %>% 
    t %>% data.frame(Date = dates)
  significances <- sapply(output, function(item) item[['Significances']] ) %>% 
    t %>% data.frame(Date = dates)
  submissions <- sapply(output, function(item) {
    c(item[['Positions']], item[['Assertions']]) %>% 
      setNames(c('Positions','Assertions'))
  }) %>% t %>% data.frame(Date = dates)
  class(freqs) <- c("freq_dates", class(freqs))
  class(submissions) <- c("submission_dates", class(submissions))
  class(significances) <- c("significance_dates", class(significances))
  if (!missing(prevalence) & !missing(case_freq)) {
    penetrance <- (prevalence * case_freq / select(freqs, -Date)) %>% 
      data.frame(Date = get_date_list())
    class(penetrance) <- c("penetrance_dates", class(penetrance))
    if (plot.it) plot(penetrance)
    return(list(freqs, submissions, significances, penetrance) %>% 
             setNames(c('Frequencies','Submissions','Significances','Penetrance')))
  }
  if (plot.it) {
    plot(freqs)
    plot(submissions)
    #plot(significances)
  }
  return(list(freqs, submissions, significances) %>% 
           setNames(c('Frequencies','Submissions','Significances')))
}


#' Plot Aggregate Population Frequencies Over Time
#'
#' This function plots the aggregate population frequency generated by freq_over_time(). 
#'    
#' @usage plot(freqs)
#' @param freqs the named frequencies output by freq_over_time(). 
#' @export

plot.freq_dates <- function(freqs) {
  freqs %>% gather(Population, Frequency, AFR, AMR, EAS, EUR, SAS, TOTAL) %>% 
    ggplot(aes(x = Date, y = Frequency, group = Population)) +
    geom_line(aes(color = Population), lwd = 1.2) + 
    geom_point(aes(fill = Population), pch = 1) + 
    ggtitle("Aggregate Population Frequency") %>% 
  return()
}

#' Plot Aggregated Penetrance Over Time
#'
#' This function plots the aggregated penetrance values 
#' generated by over_time(). 
#'    
#' @usage plot(penetrance)
#' @param penetrance the named penetrance output generated by over_time(). 
#' #@export

plot.penetrance_dates <- function(penetrance) {
  penetrance %>% gather(Population, Frequency, AFR, AMR, EAS, EUR, SAS, TOTAL) %>% 
    ggplot(aes(x = Date, y = Frequency, group = Population)) +
    geom_line(aes(color = Population), lwd = 1.2) + 
    geom_point(aes(fill = Population), pch = 1) + 
    ggtitle("Aggregate Penetrance") %>% 
    return()
}

#' Plot Submissions Over Time
#'
#' This function plots the aggregate counts of positions
#' and assertions generated by over_time(). 
#'    
#' @usage plot(submissions)
#' @param submissions the named frequencies output generated by over_time(). 
#' @export

plot.submission_dates <- function(submissions, ...) {
  submissions %>% gather(Type, Count, Positions, Assertions) %>% 
    ggplot(aes(x = Date, y = Count, group = Type)) +
    geom_line(aes(color = Type), lwd = 1.2) + 
    geom_point(aes(fill = Type), pch = 1) + 
    ggtitle("Submissions in Gene List") %>% 
  return()
}

#' Plot Significances Over Time
#'
#' This function plots the aggregate counts of variants
#' of each significances category, as generated by over_time(). 
#'    
#' @usage plot(significances)
#' @param significances the named signifance output generated by over_time(). 
#' @export

plot.significance_dates <- function(significances, ...) {
  significances %>% gather(Type, Count, Uncertain, Not.Provided, Benign, 
                           Likely.Benign, Likely.Pathogenic, Pathogenic, 
                           Drug.Response, Histocompatibility, Other) %>% 
    filter(!(Type %in% c("Histocompatibility","Other","Drug.Response"))) %>%
    ggplot(aes(x = Date, y = Count, group = Type)) +
    geom_line(aes(color = Type), lwd = 1.2) + 
    geom_point(aes(fill = Type), pch = 1) + 
    ggtitle("Significances in Gene List") %>% 
    return()
}
